sources/kafka/__init__.py

```diff
@@ -0,0 +1,91 @@
+"""A source to extract Kafka messages.
+
+When extraction starts, partitions length is checked -
+data is read only up to it, overriding the default Kafka's
+behavior of waiting for new messages in endless loop.
+"""
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+
+from confluent_kafka import Consumer, Message  # type: ignore
+
+import dlt
+from dlt.common.time import ensure_pendulum_datetime
+from dlt.common.typing import TDataItem, TAnyDateTime
+from .helpers import (
+    default_message_processor,
+    KafkaCredentials,
+    OffsetTracker,
+)
+
+
+@dlt.resource(
+    name="kafka_messages",
+    table_name=lambda msg: msg["_kafka"]["topic"],
+    standalone=True,
+)
+def kafka_consumer(
+    topics: Union[str, List[str]],
+    credentials: Union[KafkaCredentials, Consumer] = dlt.secrets.value,
+    msg_processor: Optional[
+        Callable[[Message], Dict[str, Any]]
+    ] = default_message_processor,
+    batch_size: Optional[int] = 3000,
+    start_from: Optional[TAnyDateTime] = None,
+) -> Iterable[TDataItem]:
+    """Extract recent messages from the given Kafka topics.
+
+    The resource tracks offsets for all the topics and partitions,
+    and so reads data incrementally.
+
+    Messages from different topics are saved in different tables.
+
+    Args:
+        topics (Union[str, List[str]]): Names of topics to extract.
+        credentials (Optional[Union[KafkaCredentials, Consumer]]):
+            Auth credentials or an initiated Kafka consumer. By default,
+            is taken from secrets.
+        msg_processor(Optional[Callable]): A function-converter,
+            which'll process every Kafka message after it's read and
+            before it's transfered to the destination.
+        batch_size (Optional[int]): Messages batch size to read at once.
+        start_from (Optional[TAnyDateTime]): A timestamp, at which to start
+            reading. Older messages are ignored.
+
+    Yields:
+        Iterable[TDataItem]: Kafka messages.
+    """
+    if not isinstance(topics, list):
+        topics = [topics]
+
+    if isinstance(credentials, Consumer):
+        consumer = credentials
+    elif isinstance(credentials, KafkaCredentials):
+        consumer = credentials.init_consumer()
+    else:
+        raise TypeError(
+            (
+                "Wrong credentials type provided. The credentials need to "
+                "be of type: KafkaCredentials or confluent_kafka.Consumer"
+            )
+        )
+
+    if start_from is not None:
+        start_from = ensure_pendulum_datetime(start_from)
+
+    consumer.subscribe(topics)
+    tracker = OffsetTracker(consumer, topics, dlt.current.resource_state(), start_from)
+
+    # read the messages up to the maximum offsets,
+    # not waiting for new messages
+    while tracker.has_unread:
+        batch = []
+        for msg in consumer.consume(batch_size, timeout=1):
+            if msg.error():
+                print(f"ERROR: {msg.error()}")
```

Comment:
any ideas what message error could be? we typically raise and abort on any errors but maybe it is something different? I expect the problem with the read is an exception

Reply:
Actually, yes, I'm periodically getting an error like this, when I start extraction:
```ERROR: KafkaError{code=UNKNOWN_TOPIC_OR_PART,val=3,str="Subscribed topic not available: books: Broker: Unknown topic or partition"}```

The thing is - several seconds pass, and it reads messages without problems (in the same loop). According to the docs, when you just start reading, it may take several seconds before consumer will be balanced to a broker - seems, we're trying to read before balancing finished, but it's not a reason to abort extraction.

The list of possible errors:
https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.KafkaError

There is a couple of methods, which can be used: `fatal()` and `retriable()`. If the error is not fatal or is retriable (it's not the same), then we continue. Otherwise, we'll abort extraction. Seems logical to me, will do it.

Reply:
yeah! what I do not understand in the documentation is what `fatal` means. if `fatal` is a fatal error concerning particular message - you should rather skip the message and increase the offset so you never re-read it.
if the fatal error means that the consumer should disconnect because no further message consumption is possible then we need to raise. pls make sure what is going on

Reply:
@rudolfix, this is what the docs say about fatal errors:
```Idempotent producer will typically only raise fatal errors for unrecoverable cluster errors where the idempotency guarantees can’t be maintained```

So, it's about serios failures of the whole system.

As for errors on particular messages - I don't think it's possible. By default, Kafka doesn't store messages that have problems, but raises exceptions. However, the user may create a `dead queue` to save unprocessable messages in a separate topic. In this case they'll only need to add one more topic into the resource before extraction, and with a custom message processor they'll solve whatever is wrong there.